README.md
## Features

A widget displays an encrypted image from the asset folder. The best package is to secure your local
resource. It commonly uses in some apps such as Home Workout, Math Riddles and Puzzles, English
Learning,... 

Reference:
- 📁 FileCryptor (https://github.com/webdastur/file_cryptor)

## Getting started

You should ensure that you add the router as a dependency in your flutter project.

```yaml
dependencies:
  encrypted_asset_image: ^0.0.1
```

You should then run `flutter pub get`. Don't forget to declare assets resource in `pubspec.yaml`.

```yaml
flutter:
  assets:
    - assets/
```

## Example Project

There is a example project in the `example` folder. Check it out. However, to encrypt image, you can
check out in `crypto_example` folder.

## Usage

### 1 - Generate an encrypted image
```dart
import 'dart:io';
import 'package:file_cryptor/src/file_cryptor.dart';

void main() async {
  // Init file crypto with key length = 32
  FileCryptor fileCryptor = FileCryptor(
    key: "VihW5CNfR9Fmhgz6b5AbUDQPsAzRWCA8",
    iv: 16,
    dir: "crypto_example",
  );
  const fileName = 'logo_image.png';

  // Encrypt file and calculate executed time
  var stopwatch1 = Stopwatch()..start();

  File encryptedFile = await fileCryptor.encrypt(
      inputFile: "raw/$fileName", outputFile: "encrypt/$fileName");

  print('Encryption executed in ${stopwatch1.elapsed}');
  print('Encrypted file path: ${encryptedFile.absolute}');

  // Decrypt file and calculate executed time
  var stopwatch2 = Stopwatch()..start();

  File decryptedFile = await fileCryptor.decrypt(
      inputFile: "encrypt/$fileName", outputFile: "decrypt/$fileName");

  print('Decryption executed in ${stopwatch2.elapsed}');
  print('Decrypted file path: ${decryptedFile.absolute}');
}
```
### 2 - EncryptedAssetImage

Basic widget

```dart
final FileCryptor fileCryptor = FileCryptor(
    key: 'VihW5CNfR9Fmhgz6b5AbUDQPsAzRWCA8', // This is your key with a length equal to 32
    iv: 16, // iv is Initialization vector encryption times
    dir: 'example', // Not required for this widget
    );
```
```dart
new EncryptedAssetImage(
    assetPath: 'assets/encrypted_logo_image.png',
    fileCryptor: fileCryptor)
```
Complete example

```dart
import 'package:encrypted_asset_image/encrypted_asset_image.dart';
import 'package:flutter/material.dart';

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  late final FileCryptor fileCryptor;

  @override
  void initState() {
    fileCryptor = FileCryptor(
      key: 'VihW5CNfR9Fmhgz6b5AbUDQPsAzRWCA8', // This is your key with a length equal to 32
      iv: 16, // iv is Initialization vector encryption times
      dir: '', // Not required for this widget
    );
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Encrypted Asset Image Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Expanded(
                child: EncryptedAssetImage(
                    assetPath: 'assets/encrypted_logo_image.png',
                    fileCryptor: fileCryptor)),
            Expanded(
                child: EncryptedAssetImage(
                    assetPath: 'assets/encrypted_dash_image.png',
                    fileCryptor: fileCryptor))
          ],
        ),
      ),
    );
  }
}
```

## Additional information

You can follow me on Linkedin [DuocNP] (https://www.linkedin.com/in/duocnguyen6799)

CHANGELOG.md
## 0.0.1

* Initial package

## 0.0.1+1

* Update documentation comments


lib/src/main.dart
/*
 Created by DuocNP on 06/10/2022.
 Copyright (c) 2022 . All rights reserved.
*/
import 'dart:typed_data';

import 'package:file_cryptor/file_cryptor.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class EncryptedAssetImage extends StatefulWidget {
  const EncryptedAssetImage({
    Key? key,
    required this.assetPath,
    required this.fileCryptor,
    this.scale = 1.0,
    this.frameBuilder,
    this.errorBuilder,
    this.semanticLabel,
    this.excludeFromSemantics = false,
    this.width,
    this.height,
    this.color,
    this.opacity,
    this.colorBlendMode,
    this.fit,
    this.alignment = Alignment.center,
    this.repeat = ImageRepeat.noRepeat,
    this.centerSlice,
    this.matchTextDirection = false,
    this.gaplessPlayback = false,
    this.isAntiAlias = false,
    this.filterQuality = FilterQuality.low,
    int? cacheWidth,
    int? cacheHeight,
    this.loadingBuilder,
  }) : super(key: key);

  /// Path of the asset image
  final String assetPath;

  /// Instance of FileCryptor
  final FileCryptor fileCryptor;

  /// The param of [Image.memory]
  final double scale;

  /// The param of [Image.memory]
  final ImageFrameBuilder? frameBuilder;

  /// The param of [Image.memory]
  final ImageLoadingBuilder? loadingBuilder;

  /// The param of [Image.memory]
  final ImageErrorWidgetBuilder? errorBuilder;

  /// The param of [Image.memory]
  final double? width;

  /// The param of [Image.memory]
  final double? height;

  /// The param of [Image.memory]
  final Color? color;

  /// The param of [Image.memory]
  final Animation<double>? opacity;

  /// The param of [Image.memory]
  final FilterQuality filterQuality;

  /// The param of [Image.memory]
  final BlendMode? colorBlendMode;

  /// The param of [Image.memory]
  final BoxFit? fit;

  /// The param of [Image.memory]
  final AlignmentGeometry alignment;

  /// The param of [Image.memory]
  final ImageRepeat repeat;

  /// The param of [Image.memory]
  final Rect? centerSlice;

  /// The param of [Image.memory]
  final bool matchTextDirection;

  /// The param of [Image.memory]
  final bool gaplessPlayback;

  /// The param of [Image.memory]
  final String? semanticLabel;

  /// The param of [Image.memory]
  final bool excludeFromSemantics;

  /// The param of [Image.memory]
  final bool isAntiAlias;

  @override
  State<EncryptedAssetImage> createState() => _EncryptedAssetImageState();
}

class _EncryptedAssetImageState extends State<EncryptedAssetImage> {
  FileCryptor get fileCryptor => widget.fileCryptor;

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<Uint8List>(
      future: _getEncryptedImage(assetPath: widget.assetPath),
      builder: (context, snapshot) {
        if (snapshot.hasData) {
          return Image.memory(
            snapshot.data!,
            key: widget.key,
            scale: widget.scale,
            frameBuilder: widget.frameBuilder,
            errorBuilder: widget.errorBuilder,
            semanticLabel: widget.semanticLabel,
            excludeFromSemantics: widget.excludeFromSemantics,
            width: widget.width,
            height: widget.height,
            color: widget.color,
            opacity: widget.opacity,
            colorBlendMode: widget.colorBlendMode,
            fit: widget.fit,
            alignment: widget.alignment,
            repeat: widget.repeat,
            centerSlice: widget.centerSlice,
            matchTextDirection: widget.matchTextDirection,
            gaplessPlayback: widget.gaplessPlayback,
            isAntiAlias: widget.isAntiAlias,
            filterQuality: widget.filterQuality,
          );
        } else if (snapshot.hasError) {
          return ErrorWidget('Failed to load image');
        }
        return const SizedBox();
      },
    );
  }

  /// Using to get a encrypted image from asset
  Future<Uint8List> _getEncryptedImage({required String assetPath}) async {
    ByteData bytesData = await rootBundle.load(assetPath);
    Uint8List uInt8List = bytesData.buffer
        .asUint8List(bytesData.offsetInBytes, bytesData.lengthInBytes);
    return fileCryptor.decryptUint8List(data: uInt8List);
  }
}


crypto_example/cryto_example.dart
import 'dart:io';

import 'package:file_cryptor/src/file_cryptor.dart';

void main() async {
  // Init file crypto with key length = 32
  FileCryptor fileCryptor = FileCryptor(
    key: "VihW5CNfR9Fmhgz6b5AbUDQPsAzRWCA8",
    iv: 16,
    dir: "crypto_example",
  );
  const fileName = 'logo_image.png';
  // const fileName = 'dash_image.png';

  // Encrypt file and calculate executed time
  var stopwatch1 = Stopwatch()..start();

  File encryptedFile = await fileCryptor.encrypt(
      inputFile: "raw/$fileName", outputFile: "encrypt/$fileName");

  print('Encryption executed in ${stopwatch1.elapsed}');
  print('Encrypted file path: ${encryptedFile.absolute}');

  // Decrypt file and calculate executed time
  var stopwatch2 = Stopwatch()..start();

  File decryptedFile = await fileCryptor.decrypt(
      inputFile: "encrypt/$fileName", outputFile: "decrypt/$fileName");

  print('Decryption executed in ${stopwatch2.elapsed}');
  print('Decrypted file path: ${decryptedFile.absolute}');
}
